src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -91,19 +100,72 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
             return null;
         }
 
+        // Because SemanticToken data is generated relative to the previous token it must be in order.
+        // We have a guarantee of order within any given language server, but when we translate from the ranges
+        // in the C# document to ranges into the Razor document we lose that guarantee.
+        // Having converted them to SemanticRange objects, we can simply do the final round of a merge sort.
         var newList = new List<SemanticRange>(ranges1.Count + ranges2.Count);
-        newList.AddRange(ranges1);
-        newList.AddRange(ranges2);
+        var idxRange1 = 0;
+        var idxRange2 = 0;
+        while (idxRange1 < ranges1.Count && idxRange2 < ranges2.Count)
+        {
+            var range1 = ranges1[idxRange1];
+            var range2 = ranges2[idxRange2];
 
-        // Because SemanticToken data is generated relative to the previous token it must be in order.
-        // We have a guarantee of order within any given language server, but the interweaving of them can be quite complex.
+            if (range1.CompareTo(range2) < 0)
```

Comment:
If this code is updated to handle when the two ranges are equal, and prefers `range1` in that case, I think we can therefore get rid of the entire `FromRazor` concept in the `SemanticRange` class.

It would mean the `#if DEBUG` section below, that validates the sort, would stop working, but I think we have enough semantic tokens tests to not need that in the long term.

Reply:
In #8875 I added a [`FromRazor`](https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/SemanticRange.cs#L16) property to `SemanticRange`, as there are circumstances where Razor and C# both supply a semantic token for the same piece of code (eg, a Blazor component attribute), and we wanted to always take the Razor version in those cases. I implemented it in the `CompareTo` method on `SemanticRange` so that the sorting algorithm would put things in the right order, and then [added code](https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs#L388-L393) to skip overlapping tokens when converting the `SemanticRange` objects back to the 5-int deltas.

It seems like both of those things could be simplified and moved into this method, now that the sort algorithm isn't a black box `.Sort()` method.

I also think this is totally fine to do as a follow up if you prefer, and would happily do the work, since its my code it would be removing.😁

Reply:
I've finished that integration.